### PR TITLE
feat(mock): add comprehensive mock logger handler for testing

### PR DESCRIPTION
**Pull request by amery**
### **User description**
# feat(mock): Add comprehensive mock logger handler for testing

## Summary

This PR introduces a new **mock logger handler** that provides a complete,
thread-safe implementation of `slog.Logger` for testing and verification
purposes. The mock handler records all log messages with their metadata
(levels, fields, stack traces) for programmatic inspection, making it ideal for
testing applications and other slog handlers.

### 🎯 **Complete slog.Logger Implementation**

- Full implementation of all `slog.Logger` interface methods
- Proper handling of log levels (Debug, Info, Warn, Error, Fatal, Panic)
- Support for structured fields via `WithField()` and `WithFields()`
- Stack trace recording with `WithStack()`
- Immutable logger behaviour with proper method chaining

### 🔒 **Thread-Safe Architecture**

- Concurrent-safe message recording using mutex protection
- Safe for use across multiple goroutines
- No race conditions in message capture or retrieval

### 📊 **Comprehensive Message Capture**

```go
type Message struct {
    Message string         // The logged message text
    Level   slog.LogLevel  // Log level (Debug, Info, etc.)
    Fields  map[string]any // All attached structured fields
    Stack   bool           // Whether stack trace was requested
}
```

### 🧪 **Testing-Focused Design**

- **Logger**: Primary interface for applications under test
- **Recorder**: Separate storage component for advanced testing scenarios
- **Message**: Rich data structure capturing all log metadata
- Formatted string output with `Message.String()` for easy debugging

### Core Components

```go
// Create a mock logger
logger := mock.NewLogger()

// Use it like any slog.Logger
logger.Info().
    WithField("user", "john").
    WithField("action", "login").
    Print("User authentication successful")

// Retrieve and verify logged messages
messages := logger.GetMessages()
```

### Advanced Usage

```go
// Create separate recorder for shared testing
recorder := mock.NewRecorder()

// Direct message recording
recorder.Record(mock.Message{
    Message: "Custom message",
    Level:   slog.Error,
    Fields:  map[string]any{"error": "something failed"},
})
```

### 🔍 **Application Testing**

Perfect for testing application logic that uses logging:

```go
func TestUserService(t *testing.T) {
    logger := mock.NewLogger()
    service := NewUserService(logger)

    err := service.CreateUser("john@example.com")

    messages := logger.GetMessages()
    // Verify appropriate logging occurred
}
```

### 🔌 **Handler Adapter Testing**

Ideal for testing custom slog handlers:

```go
func TestMyHandler(t *testing.T) {
    backend := mock.NewLogger()
    handler := myhandler.New(backend)

    handler.Error().Print("test error")

    // Verify handler correctly forwarded to backend
    messages := backend.GetMessages()
    assert.Equal(t, slog.Error, messages[0].Level)
}
```

### 🏗️ **Immutable Logger Architecture**

The mock logger maintains immutability using `internal.Loglet` for consistent
field chain management like other slog handlers:

```go
func (l *Logger) WithField(label string, value any) slog.Logger {
    if l == nil {
        return nil
    }
    if label == "" {
        return l  // Same instance for empty keys
    }
    return l.withLoglet(l.Loglet.WithField(label, value))
}

func (l *Logger) withLoglet(loglet internal.Loglet) *Logger {
    return &Logger{
        Loglet:   loglet,    // Immutable field chain
        recorder: l.recorder, // Shared message storage
        enabled:  l.enabled,
    }
}
```

### 🔐 **Nil-Safe Operations**

All methods handle nil receivers gracefully:

```go
func (l *Logger) GetMessages() []Message {
    if l == nil {
        return nil
    }
    return l.recorder.GetMessages()
}
```

### 📚 **Internal Testing Migration**

The existing `internal/testing` package now provides backward-compatible
aliases:

```go
// Old approach (still supported)
import slogtest "darvaza.org/slog/internal/testing"
logger := slogtest.NewLogger()

// New recommended approach
import "darvaza.org/slog/handlers/mock"
logger := mock.NewLogger()
```

### ✅ **Comprehensive Test Suite**

- **89.3% code coverage** with exhaustive edge case testing
- Table-driven tests following project conventions
- Nil receiver testing for robustness
- Concurrent access testing for thread safety
- Field immutability and isolation verification

### 📏 **Code Quality Standards**

- Follows project linting requirements (complexity limits, formatting)
- Proper use of `t.Helper()` in test utilities
- Safe type assertions with error handling
- Comprehensive documentation and examples

## Files Changed

```diff
README.md                    |  31 +++
handlers/mock/README.md      | 136 ++++++++++++
handlers/mock/mock.go        | 303 ++++++++++++++++++++++++++
handlers/mock/mock_test.go   | 509 +++++++++++++++++++++++++++++++++++++++++++
internal/testing/README.md   |  96 +++++++-
internal/testing/helpers.go  |   5 -
internal/testing/recorder.go | 216 ++----------------
7 files changed, 1088 insertions(+), 208 deletions(-)
```

### ✨ **Enhanced Testing Capabilities**

- Provides missing piece for comprehensive slog testing
- Enables verification of logging behaviour in applications
- Simplifies testing of custom handlers and adapters

### 🏗️ **Better Architecture**

- Moves testing utilities from internal to public API
- Follows Go best practices for testing infrastructure
- Maintains backward compatibility while encouraging modern usage

This mock handler completes the slog ecosystem by providing essential testing
infrastructure as a first-class public API, enabling robust testing of
applications and handlers that use structured logging.

___

### **PR Type**
Enhancement

___

### **Description**
- Add public mock logger handler for testing slog applications

- Move internal testing utilities to handlers/mock package

- Provide backward compatibility aliases in internal/testing

- Include comprehensive documentation and examples

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["internal/testing"] --> B["handlers/mock"]
  B --> C["Public API"]
  A --> D["Backward Compatibility"]
  C --> E["Testing Applications"]
  C --> F["Handler Development"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mock.go</strong><dd><code>Complete mock logger implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

handlers/mock/mock.go

<ul><li>Implement complete mock logger with Message, Recorder, and Logger <br>types<br> <li> Provide thread-safe message recording with mutex protection<br> <li> Support all slog.Logger interface methods with immutable behavior<br> <li> Include nil-safe operations and proper field copying</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/126/files#diff-625a2b79a7f6e6d4a990140e15406d839ba6717897a12265523511553b370338">+303/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mock_test.go</strong><dd><code>Comprehensive mock logger test suite</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

handlers/mock/mock_test.go

<ul><li>Add comprehensive test suite with 89.3% code coverage<br> <li> Test all logger methods, field handling, and thread safety<br> <li> Include nil receiver tests and immutability verification<br> <li> Use table-driven tests with helper functions</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/126/files#diff-54a4858c879e60a28e67a0ee9d8769be48a5ce63a4b1c50227d1b6333096804d">+509/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Miscellaneous</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>helpers.go</strong><dd><code>Remove moved Clear method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/testing/helpers.go

- Remove Clear method that was moved to mock package

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/126/files#diff-04219bd2ccfc98e150359828693a30148d71e5bcd2802a02491f6c9d1417d08e">+0/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>recorder.go</strong><dd><code>Backward compatibility aliases for moved types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/testing/recorder.go

<ul><li>Replace implementation with backward compatibility aliases<br> <li> Provide deprecated type aliases for Message, Recorder, Logger<br> <li> Maintain existing API while directing to public mock package</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/126/files#diff-736df2e7aa33f1278a15392a4642ec669d1b8061fe252435daa9c0d5e6d661e1">+17/-199</a></td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Document mock handler in main README</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

README.md

<ul><li>Add mock handler to utility handlers section<br> <li> Include usage example for testing applications<br> <li> Document message verification patterns</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/126/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+31/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Complete mock handler documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

handlers/mock/README.md

<ul><li>Comprehensive documentation with usage examples<br> <li> Cover basic testing, field handling, and handler adapter testing<br> <li> Document API reference and design notes</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/126/files#diff-5e0c9464d4f641c696ce2595132a257378176e6d750eb434d1f841a4e25b341c">+136/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Migration documentation for public API</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/testing/README.md

<ul><li>Add migration guidance to public mock package<br> <li> Update examples showing both old and new approaches<br> <li> Document backward compatibility and migration steps</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/126/files#diff-cdf4dc66b07b31a0ed9e3325f6a9df3806938c605ea2855ea00054d759f51b4b">+92/-4</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

* **New Features**
  * Introduced a mock logger for capturing and inspecting log messages during testing, with full documentation and usage examples.
* **Documentation**
  * Added and updated README files to document the new mock logger, its API, and migration instructions for users transitioning from internal to public packages.
* **Tests**
  * Added comprehensive unit tests for the mock logger, ensuring correct logging behavior, field handling, immutability, and thread safety.
* **Refactor**
  * Migrated internal testing utilities to use the new public mock logger package, maintaining backward compatibility with type aliases and updated documentation.
* **Chores**
  * Removed deprecated or redundant internal implementations, streamlining the codebase and encouraging use of the public API.

### CODE REVIEW

**Comment by amery** [comment]
## Implementation Update

Refactored mock.Logger to use `internal.Loglet` for field management instead of custom field copying system.

**Key changes:**
- Embed `internal.Loglet` in Logger struct for consistent field chain behavior
- Replace custom `copyFields()` with standard `withLoglet()` helper pattern  
- Add proper empty key/field checks like other handlers (filter, logrus, etc.)
- Maintain same-instance returns for no-op operations

This ensures the mock handler behaves identically to other slog handlers while preserving all testing functionality. All tests pass with same coverage.

File: handlers/mock/mock.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+	return l.enabled
+}
+
+// WithEnabled returns the logger and its enabled state.
+func (l *Logger) WithEnabled() (slog.Logger, bool) {
+	if l == nil {
+		return nil, false
+	}
+	return l, l.enabled
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply handlers/mock/mock.go]: excellent point, it should be sorted now
